### PR TITLE
provide `SeekableStream` impl for `tokio::fs::File`

### PR DESCRIPTION
**Pull request by demoray**
Addresses #1219

This builds upon #1359, #1358, and #1357 and those are required for this PR to build correctly.

### CODE REVIEW

File: sdk/core/src/tokio/fs.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        // #[doc = "Amount of data to read from the file"]
+        block_size: u64 => Some(block_size),
+        // #[doc = "Amount of data to buffer in memory during streaming reads"]
+        buffer_size: usize => Some(buffer_size),
+    }
+
+    pub async fn build(mut self) -> crate::Result<FileStream> {
+        let stream_size = self.handle.metadata().await?.len();
+
+        let buffer_size = self.buffer_size.unwrap_or(1024 * 64);
```
Comment (johnbatty) [inline sdk/core/src/tokio/fs.rs:43]: suggestion: Use a named constant for the default buffer_size

File: sdk/core/src/tokio/fs.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+impl FileStream {
+    async fn read(&mut self, slice: &mut [u8]) -> std::io::Result<usize> {
+        let mut handle = self.handle.clone().lock_owned().await;
+        handle.read(slice).await
+    }
+
+    /// Resets the number of bytes that will be read from this instance to the
+    /// `stream_size`
+    ///
+    /// This is useful if you want to read the stream in mutliple blocks
```
Comment (johnbatty) [inline sdk/core/src/tokio/fs.rs:90]: typo: "mutliple" => "multiple"

File: sdk/core/src/tokio/fs.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        );
+        min(self.stream_size - self.offset, self.block_size) as usize
+    }
+
+    /*
+    fn buffer_size(&self) -> usize {
+        self.buffer_size
+    }
+     */
+}
```
Comment (yoshuawuyts) [inline sdk/core/src/tokio/fs.rs:143]: Did you intend to leave this comment in?

File: sdk/core/src/tokio/fs.rs [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    #[pin]
+    handle: Arc<Mutex<Take<File>>>,
+    pub stream_size: u64,
+    pub block_size: u64,
+    buffer_size: usize,
+    pub offset: u64,
+}
+
+impl FileStream {
+    async fn read(&mut self, slice: &mut [u8]) -> std::io::Result<usize> {
```
Comment (yoshuawuyts) [inline sdk/core/src/tokio/fs.rs:86]: What is the difference between this inherent method and `AsyncRead::read`? A comment explaining that here would probably be helpful.

**Review by yoshuawuyts — APPROVED** [review]
Some minor nits, but overall I think this looks good!

File: sdk/core/src/tokio/fs.rs (hunk 3 above)
Comment (demoray) [reply sdk/core/src/tokio/fs.rs]: This is an artifact from trying to split up the work into multiple PRs.  In this case, #1359.

File: sdk/core/src/tokio/fs.rs (hunk 4 above)
Comment (demoray) [reply sdk/core/src/tokio/fs.rs:92]: Added.

**Pull request merged**